client/client.go

```diff
@@ -122,25 +123,46 @@ func bind(cfg *Config, conn ldaputil.Connection) error {
 		return errors.New("unable to bind due to lack of configured password")
 	}
 
-	if cfg.BindDN == "" {
-		return errors.New("must provide binddn")
+	// Determine the user to bind with
+	var bindUser string
+	switch {
+	case cfg.UPNDomain != "":
+		bindUser = fmt.Sprintf("%s@%s", ldaputil.EscapeLDAPValue(cfg.BindDN), cfg.UPNDomain)
+	case cfg.BindDN != "":
+		bindUser = cfg.BindDN
+	default:
+		return errors.New("must provide binddn or upndomain")
 	}
```

Comment:
Nice! This is so much better, I hadn't noticed the fact this was the only difference between the code paths anyway.